cdc/generation.cc

```diff
@@ -720,6 +726,14 @@ future<> generation_service::check_and_repair_cdc_streams() {
         throw std::runtime_error("check_and_repair_cdc_streams: node not initialized yet");
     }
 
+    if (!_feature_service.cluster_supports_cdc_generations_v2()) {
+        throw std::runtime_error(
+                "check_and_repair_cdc_streams: we've detected a partially upgraded cluster: the CDC_GENERATIONS_V2 feature is not enabled."
+                " Please finish the rolling upgrade procedure first and then try executing the command again."
+                " If you really must execute the command right now, execute it on one of the nodes that have not yet been upgraded"
+                " (the node trying to execute the command now is upgraded), but it is highly recommended to finish the upgrade first.");
+    }
+
```

Comment:
This introduces sad coupling. Consider a Kubernetes autoscaler that adds nodes autonomously. It doesn't know if an upgrade is in progress. We could require it to know, but that propagates internal constraints to the ecosystem.

Why is this necessary? We could support both the old and new ways for a while, no?

Reply:
I wasn't aware that we "officially" support scaling cluster in the middle of an upgrade - and I'm not talking about CDC here, I'm guessing many things could go wrong.

Anyway:
1. This command should be executed VERY rarily.
2. This command is not for scaling clusters. It's for exceptional situations: recovery from fatal errors. Thus it should generally be run in controlled environment manually by an admin. One exception where it would be "commonly" run is when users upgrade between experimental versions.
3. As the message says, you can still execute this command on a non-upgraded node.
4. Executing the command on an upgraded node would introduce a generation with the new ID format, causing old nodes to get confused and log a lot of errors. If the old node doesn't have the `generation_service` refactor, i.e. the generation handling code is still in `storage_service`, this would even cause the node to stop handling status changes of other nodes which would be very bad.
5. To fix 4. I could modify the code so that if you execute the command on an upgraded node but a partially upgraded cluster, it would introduce a generation with the old ID format (using the old tables). But this would require readding back a lot of code/keeping two versions of the same code which would be a waste, there really is no value here: the admin can always run the command on an old node (see 3.) and it should not be executed by an automatic process (see 2.).
6. This check doesn't block scaling, you can still add a node to a partially upgraded cluster (although due to the check in `bootstrap()` it would have to be an old node).

Reply:
> I wasn't aware that we "officially" support scaling cluster in the middle of an upgrade - and I'm not talking about CDC here, I'm guessing many things could go wrong.

We don't support it "officiall", but the code is written to support it against the day we do. I believe it is inevitable.

> 
> Anyway:
> 
> 1. This command should be executed VERY rarily.
> 2. This command is not for scaling clusters. It's for exceptional situations: recovery from fatal errors. Thus it should generally be run in controlled environment manually by an admin. One exception where it would be "commonly" run is when users upgrade between experimental versions.
> 3. As the message says, you can still execute this command on a non-upgraded node.
> 4. Executing the command on an upgraded node would introduce a generation with the new ID format, causing old nodes to get confused and log a lot of errors. If the old node doesn't have the `generation_service` refactor, i.e. the generation handling code is still in `storage_service`, this would even cause the node to stop handling status changes of other nodes which would be very bad.
> 5. To fix 4. I could modify the code so that if you execute the command on an upgraded node but a partially upgraded cluster, it would introduce a generation with the old ID format (using the old tables). But this would require readding back a lot of code/keeping two versions of the same code which would be a waste, there really is no value here: the admin can always run the command on an old node (see 3.) and it should not be executed by an automatic process (see 2.).
> 6. This check doesn't block scaling, you can still add a node to a partially upgraded cluster (although due to the check in `bootstrap()` it would have to be an old node).

Ok, with such a list of caveats we can make an exception.

Reply:
> We don't support it "officially", but the code is written to support it against the day we do.

I don't think that matters much if we never test these scenarios (quick glance at SCT didn't reveal tests which scale in the middle of upgrade).